waku/v2/utils/pagination.nim

```diff
@@ -14,26 +14,47 @@ export hash
 type
   Index* = object
     ## This type contains the  description of an Index used in the pagination of WakuMessages
-    digest*: MDigest[256]
+    digest*: MDigest[256] # calculated over payload and content topic
     receiverTime*: Timestamp
     senderTime*: Timestamp # the time at which the message is generated
+    pubsubTopic*: string
 
 proc `==`*(x, y: Index): bool =
-  ## receiverTime plays no role in index comparison
-  (x.senderTime == y.senderTime) and (x.digest == y.digest)
+  ## receiverTime plays no role in index equality
+  (x.senderTime == y.senderTime) and
+  (x.digest == y.digest) and
+  (x.pubsubTopic == y.pubsubTopic)
 
 proc cmp*(x, y: Index): int =
   ## compares x and y
   ## returns 0 if they are equal 
   ## returns -1 if x < y
   ## returns 1 if x > y
-  ## receiverTime plays no role in index comparison
+  ## 
+  ## Default sorting order priority is:
+  ## 1. senderTimestamp
+  ## 2. receiverTimestamp (a fallback only if senderTimestamp unset on either side, and all other fields unequal)
```

Comment:
What's the rationale for not breaking ties on senderTimestamp with receiverTimestamp? I.e. if you receive two messages in same second. Seems like the special casing makes it more complicated, but maybe I'm missing something here?

Reply:
The same message on different store nodes will have slightly differing `receiverTimestamp`s. If a store node tries to synchronise/resume from multiple other store nodes, it won't be able to deduplicate any messages if it takes `receiverTimestamp` into consideration. Something similar could happen during normal `relay` if different traversal paths for the same message has enough latency diff to fall outside the underlying gossipsub deduplication buffer - i.e. two identical messages with the exact same `senderTimestamp` should be considered duplicates. Of course, this would mean that `store` duplicate detection on messages without `senderTimestamp` may not be 100% accurate, but I think this is our best option.